traitlets/traitlets.py

```diff
@@ -644,6 +648,17 @@ def hold(*a):
                 for v in cache.values():
                     self._notify_trait(*v)
 
+    @contextlib.contextmanager
+    def _allow_write(self):
```

Comment:
I don't quite understand what you mean regarding the `on_change` handler. Why would it be unlocked?
Regarding the multi-threading that you are concerned about, the way I was thinking of doing this is by scheduling a thread-safe set of the value in the zmq event loop. what would be the issue then exactly?